Dynamic Cluster Pool

gandro opened the pull request:
This adds a new IPAM mode, named `cluster-pool-v2`, where the operator is able to assign and remove additional pod CIDRs to a node based on demand. The current usage of pod CIDRs is indicated by the Cilium Agent in the local CiliumNode CRD:

```
This extends the CiliumNode CRD with two spec and one status fields
needed to support dynamic allocation of pod CIDRs to node.

  - `CiliumNode.Status.IPAM.UsedPodCIDRs` informs the operator about the
    pod CIDR usage of this node.  Each pod CIDR can be in one of three
    states:

    * `in-use`: This pod CIDR is currently being used for allocation of
      pod IPs.
    * `depleted`: Every IP in this pod CIDR is currently used up. If all
      pod CIDRs are depleted, the operator will allocate and assign new
      pod CIDRs to this node.
    * `released`: This pod CIDR is no longer in use by the agent and
      should be deallocated by the operator.

  - `CiliumNode.Spec.IPAM.PodCIDRAllocationThreshold` defines the
    minimum number of free IPs which must be available to this node via
    its pod CIDR pool. If the total number of IP addresses in the pod CIDR
    pool is less than this value, the pod CIDRs currently in-use by this
    node will be marked as depleted and cilium-operator will allocate a new
    pod CIDR to this node.  This value effectively defines the buffer of IP
    addresses available immediately without requiring cilium-operator to get
    involved.

  - `CiliumNode.Spec.IPAM.PodCIDRReleaseThreshold` defines the maximum
    number of free IPs which may be available to this node via its pod
    CIDR pool. While the total number of free IP addresses in the pod CIDR
    pool is larger than this value, cilium-agent will attempt to release
    currently unused pod CIDRs.
```

See individual commit messages for more details.

This version of the feature is in beta state. A follow-up PR will add
- Documentation
- CI test(s) to complement the existing unit test

*Please indicate in the review what feedback is a blocker for rc0, and which feedback may be addressed in a follow-up PR.*

```release-note
Support dynamic allocation of pod CIDRs in cluster pool v2 IPAM mode
```

gandro committed:
crd: Add cluster-pool-v2 fields to CiliumNode

This extends the CiliumNode CRD with two spec and one status fields
needed to support dynamic allocation of pod CIDRs to node.

  - `CiliumNode.Status.IPAM.UsedPodCIDRs` informs the operator about the
    pod CIDR usage of this node.  Each pod CIDR can be in one of three
    states:

    * `in-use`: This pod CIDR is currently being used for allocation of
      pod IPs.
    * `depleted`: Every IP in this pod CIDR is currently used up. If all
      pod CIDRs are depleted, the operator will allocate and assign new
      pod CIDRs to this node.
    * `released`: This pod CIDR is no longer in use by the agent and
      should be deallocated by the operator.

  - `CiliumNode.Spec.IPAM.PodCIDRAllocationThreshold` defines the
    minimum number of free IPs which must be available to this node via
    its pod CIDR pool. If the total number of IP addresses in the pod CIDR
    pool is less than this value, the pod CIDRs currently in-use by this
    node will be marked as depleted and cilium-operator will allocate a new
    pod CIDR to this node.  This value effectively defines the buffer of IP
    addresses available immediately without requiring cilium-operator to get
    involved.

  - `CiliumNode.Spec.IPAM.PodCIDRReleaseThreshold` defines the maximum
    number of free IPs which may be available to this node via its pod
    CIDR pool. While the total number of free IP addresses in the pod CIDR
    pool is larger than this value, cilium-agent will attempt to release
    currently unused pod CIDRs.

Co-authored-by: Tom Payne <tom@isovalent.com>
Signed-off-by: Tom Payne <tom@isovalent.com>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
cidr: Add RemoveAll helper

This commit adds a function which removes all matching net.IPNet from a
slice.

Co-authored-by: Tom Payne <tom@isovalent.com>
Signed-off-by: Tom Payne <tom@isovalent.com>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
ipam: Add cluster-pool-v2 agent implementation

This adds the agent IPAM logic for cluster-pool-v2. The module behaves
similarly to the existing CRD IPAM modes, but has a different internal
logic as it deals with pod CIDRs instead of pod IPs.

Like the other CRD IPAM modes, it creates its own CiliumNode watcher to
receive updates from the operator and updates the status field to
reflect the current pod CIDR usage.

Co-authored-by: Tom Payne <tom@isovalent.com>
Signed-off-by: Tom Payne <tom@isovalent.com>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
ipam: Add cluster-pool-v2 operator implementation

This adds the operator side of cluster-pool-v2 IPAM mode. It shares the
majority of the code with the existing cluster-pool implementation, but
as it is not fully backwards-compatible in the case of allocation
errors, it branches out into a different allocation logic.

If these backwards incompatibilities can be resolved or mitigated, the
goal is to eventually deprecate cluster-pool v1 and always default to
v2.

Co-authored-by: Tom Payne <tom@isovalent.com>
Signed-off-by: Tom Payne <tom@isovalent.com>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
health: Move endpoint IP to node package

The Cilium health endpoint IP is the only static local node field which
is not accessible from the node package. To avoid code modifying the
`NodeDiscovery.LocalNode` field (which will become more dynamic in a
subsequent mode), the health IP is now moved to the node package with a
corresponding getter and setter like all other static node fields.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
nodediscovery: Unify source of truth for node name

Clustermesh related code is relying on
`NodeDiscovery.LocalNode.FullName()` to obtain the globally unique name
of the local node. This field is initialized from `nodeTypes.GetName()`.
This code removes all uses of `LocalNode.FullName()` and replaces it
with an equivalent call to `nodeTypes.GetAbsoluteNodeName()`. This
allows us to make the `NodeDiscovery.LocalNode` private in a subsequent
commit.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
nodediscovery: Make LocalNode object private

The LocalNode currently consists of static fields which are never
changed after the local node has been initialized. However, with the
introduction of secondary allocation CIDRs, we want to be update the
secondary allocation CIDR fields dynamically at run-time.

To ensure writes to the object are always protected by a mutex, and that
to be able to ensure that any changes to the object are propagated to
relevant subsystems, this commit makes the `LocalNode` object private.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
k8s/watchers: Hide subscriber interface

This commit introduces helper methods to register subscribers to the
CiliumNode or Node chain. This makes it easier to invoke those methods
via interface in a subsequent commit.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
watchers: Propagate all local CiliumNode updates

This commit informs CiliumNode subscribers about the newest resource
version of the CiliumNode CRD. This is required to be able to perform
updates to the local CiliumNode CRD without having to fetch the lastest
version manually. A subsequent commit will make use of this.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
nodediscovery: Populate clusterpool-v2 configuration

This populates the IPAM configuration section in newly created
CiliumNode objects from the CNI configuration file. This mechanism works
very similarly to e.g. ENI IPAM mode.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
daemon: Validate IPAM cluster-pool-v2 configuration

IPSec and tunnel mode are currently not supported, as they assume a
single pod CIDR per node.

IPSec requires some work to either support multiple pod CIDRs or we need
to enable the pod subnet option (used on ENI).

Tunnel mode currently still creates a tunnel BPF map with one pod CIDR
per node. We might be able to exclusively use the IPCache tunnel
endpoint field instead, but this has not been validated yet.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

twpayne commented on pkg/ipam/clusterpool.go:
Do we still need the special case handling of the first pod CIDR? IIRC at one point we decided it was unnecessary.

gandro replied:
Yes, you are absolutely right. You reviewed it a second too soon. I have already removed it locally, but haven't pushed yet because I was waiting on CI to finish. Will push an updated version soon.

twpayne commented on pkg/ipam/clusterpool.go:
`pkg/k8s.IsLocalCiliumNode` could be used instead of this function.

twpayne commented on pkg/ipam/allocator/podcidr/podcidr_v2.go:
```diff
@@ -0,0 +1,318 @@
+	return false
+}
+
+type aggregatedErrors []error
```
Consider using `go.uber.org/multierr` (already indirectly imported) instead.

gandro replied:
Great suggestion, looks much cleaner now!

gandro committed:
node: Add field for secondary allocation CIDRs

This commit adds a new secondary allocation CIDRs (pod CIDRs) to the
node types, to support multiple pod CIDRs per node. The secondary CIDRs
will be integrated into the datapath and IPAM code in subsequent
commits.

The code distinguishes between primary and secondary pod CIDRs: The
primary pod CIDR is the CIDR from with the router and health IPs are
allocated from, thus it can never be removed from node while cilium-agent
is running. Secondary CIDRs on the other hand will be added and removed
dynamically.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
datapath: Add routes for secondary allocation CIDRs

This commit extends the datapath routing setup code to support multiple
allocation CIDRs per node, namely for local node routes (i.e. routes to
`cilium_host`) and for auto direct routes.

This commit does not update the IPSec node routes or the routes added to
the tunnel map, as this requires further changes to the code. This means
that IPSec and tunnel mode currently do not support secondary allocation
CIDRs.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
option: Add cluster-pool-v2 IPAM mode string

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

christarazi commented on pkg/datapath/linux/node.go:
```diff
@@ -250,47 +250,51 @@ func installDirectRoute(CIDR *cidr.CIDR, nodeIP net.IP) (routeSpec *netlink.Rout
 }
 
-func (n *linuxNodeHandler) updateDirectRoute(oldCIDR, newCIDR *cidr.CIDR, oldIP, newIP net.IP, firstAddition, directRouteEnabled bool) error {
+func (n *linuxNodeHandler) updateDirectRoutes(oldCIDRs, newCIDRs []*cidr.CIDR, oldIP, newIP net.IP, firstAddition, directRouteEnabled bool) error {
```
I reviewed the logic in this function and it LGTM in terms of logical gaps from the old implementation. But to be really sure, I noticed that we are lacking some unit tests for this specific function so it might be worth writing unit tests _against the old_ version first, then apply this commit and see if anything breaks (of course there will be slight modifications required as the func params changed, etc).

gandro replied:
Thanks! That's a really good suggestion. Will implement asap.

gandro replied:
Turns out, we already have a pretty good unit test for the function: https://github.com/cilium/cilium/blob/6237e74b4f0ffbfc38b5c26f5a0a8f94f33a6c69/pkg/datapath/linux/node_linux_test.go#L646

Doesn't seem like it regressed (run-time tests passed before). But I'll extend it to make sure we cover all corner cases and add support the new behavior.

gandro replied:
Extended the unit tests accordingly: https://github.com/cilium/cilium/pull/18887/commits/d646e1a17356d08f14e57e48da10646187510f9f

christarazi reviewed (commented):
As discussed offline, the overall PR looks good as a first go. My main focus was to look at the first couple of commits to identify any changes that may have caused regressions as the rest of the changes are guarded by the fact that the user must opt into running the cluster-pool-v2 mode, so the risk is mitigated. A look at the latter commits especially the last couple is deferred for now. 

With that said, I did leave a comment re: the second commit.

gandro committed:
datapath: Extend TestNodeUpdateDirectRouting

This commit extends the TestNodeUpdateDirectRouting unit test to test
for nodes with multiple allocation CIDRs.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

christarazi reviewed (approved):
:rocket:

aanm commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -58,13 +58,21 @@ func (k *K8sWatcher) ciliumNodeInit(ciliumNPClient *k8s.K8sCiliumClient, asyncCo
 							if oldCN.DeepEqual(ciliumNode) {
 								equal = true
-								return
+								if !isLocal {
+									// For remote nodes, we return early here to avoid unnecessary update events if
+									// nothing in the spec or status has changed. But for local nodes, we want to
+									// propagate the new resource version (not compared in DeepEqual) such that any
+									// CiliumNodeChain subscribers are able to perform updates to the local CiliumNode
+									// object using the most recent resource version.
+									return
+								}
```
Why aren't we comparing the local node equalness to avoid consuming unnecessary CPU cycles?

gandro replied:
You mean checking for full equality using `reflect.DeepEqual`? I can add that, depending on the subscribers it might not save us any cycles though.

aanm commented on pkg/defaults/defaults.go:
```diff
@@ -344,6 +344,14 @@ const (
 
+	// IPAMPodCIDRAllocationThreshold is the default value for
+	// CiliumNode.Spec.IPAM.PodCIDRAllocationThreshold if no value is set
+	IPAMPodCIDRAllocationThreshold = 8
```
Might be worth just adding a small comment here on why 8 was picked.

aanm commented on pkg/defaults/defaults.go:
```diff
@@ -344,6 +344,14 @@ const (
+
+	// IPAMPodCIDRReleaseThreshold is the default value for
+	// CiliumNode.Spec.IPAM.PodCIDRReleaseThreshold if no value is set
+	IPAMPodCIDRReleaseThreshold = 16
```
Might be worth just adding a small comment here on why 16 was picked.

aanm commented on pkg/ipam/types/types.go:
```diff
@@ -115,6 +134,11 @@ type IPAMStatus struct {
+	// UsedPodCIDRs lists the status of each pod CIDR allocated to this node.
+	//
+	// +optional
+	UsedPodCIDRs UsedPodCIDRMap `json:"used-pod-cidrs,omitempty"`
```
The `used-pod-cidr` could be renamed to `pod-cidr` because we have a field status for each pod-cidr. Having `used` in here might be misleading if a pod cidr was "`released`: This pod CIDR is no longer in use by the agent and should be deallocated by the operator.", which begs the question is this podCIDR "used" or "released"?

aanm commented on pkg/ipam/option/option.go:
```diff
@@ -23,6 +23,9 @@ const (
 	IPAMClusterPool = "cluster-pool"
 
+	// IPAMClusterPoolV2 is the value to select cluster pool version 2
+	IPAMClusterPoolV2 = "cluster-pool-v2"
```
`cluster-pool-v2` sounds good for now, we can deprecate the old behavior of `cluster-pool` if `cluster-pool-v2` is solid for the v1.12 release

gandro replied:
Yes, this is my thinking as well. There are still some known shortcomings with v2, such as the lack of IPSec support that need to be addressed first.

aanm commented on pkg/ipam/clusterpool.go:
Not sure, maybe it is worth double check that `allocationThreshold` is never less than `releaseThreshold`?

aanm reviewed (approved):
Awesome work :tada: Only some comments, none of them are blocking and can be addressed afterwards.

joestringer commented on pkg/node/address.go:
```diff
@@ -688,6 +692,26 @@ func GetWireguardPubKey() string {
+	ipv6HealthIP = ip
+}
+
+// GetEndpointHealthIPv6 returns the IPv6 cilium-health endpoint address.
+func GetEndpointHealthIPv6() net.IP {
+	return ipv6HealthIP
+}
```
I know that this style is more consistent with the way that most of `pkg/node` works, but I was hoping not to expand on this model. We used this style quite a bit across the Cilium codebase, and it's often one of the reasons that unit testing is difficult, because you can't just declare an interface & mock/dependency-inject such code.

Looking at the caller, previously the dependency here was explicitly passed into the initialization function, but now it calls this getter and the ordering is less explicit around how this value gets populated.  I think it's harder to trace both (1) initialization + (2) updates to see we are correctly handling the updates, especially now that this is more dynamic.

Have you thought about this too? I don't think this is appropriate to have in the discussion around this particular PR, so I don't want to block this. I would like to start a thread of discussion on some code maintenance / code debt items that I see like this though. At least hopefully get on the same page on whether we think it continues to make sense to accept code in this style or not.

gandro replied:
Yes, this also came up during the review session with Chris, Michi and Tom. I fully agree that we should get rid of the `pkg/node` pattern. Besides the problems you mentioned, we also we have two sources of truth for the local node (`pkg/node` and `pkg/nodediscovery`) which I think is a potential footgun.

My thinking here is also that we should get rid of the `pkg/node` getters. But I didn't want to tackle that refactor in this PR (both due to time constraints and the size of the PR). The sole reason I added getters to `pkg/node` instead of `pkg/nodediscovery.NodeDiscovery` is that at its more consistent with the existing code.

Let's create an issue or thread discussing how to address this technical debt in a follow-up PR.

joestringer reviewed (approved):
I mostly just looked at health since the "datapath" (userspace node handling) bits were already reviewed by Chris. Just one non-blocking comment.

rolinh reviewed (approved):
Due to the sheer size of the PR, I limited my review to the topic I was pulled in for: check vendor changes. Turned the review from a lengthy one to a 2s one :grin: 

PS: I like seeing so many tests :heart_eyes:

kaworu commented on pkg/ipam/allocator/podcidr/podcidr.go:
```diff
@@ -358,13 +360,36 @@ func (n *NodesPodCIDRManager) Update(node *v2.CiliumNode) bool {
+		// if allocated is false it means that we were unable to allocate
+		// a CIDR so we need to update the status of the node into k8s.
+		updateStatus = !allocated && updateStatus
+		// ClusterPool v1 never both the spec and the status
```
typo?
```
 		// ClusterPool v1 never returns both the spec and the status
```

gandro replied:
Yes, thanks!

kaworu commented on pkg/cidr/cidr_test.go:
```diff
@@ -189,3 +189,112 @@ func (t *CidrTestSuite) TestEqual(c *check.C) {
+	}
+	for _, tt := range tests {
+		result := RemoveAll(tt.args.ipNets, tt.args.toRemove)
+		c.Assert(result, checker.DeepEquals, tt.want, check.Commentf("Test Name: %s", tt.name))
```
nitpicking: as I understand it `RemoveAll` modify its first slice parameter by design, and if correct then I think this behaviour should be tested as well.

>  		c.Assert(tt.args.ipNets, checker.DeepEquals, tt.want, check.Commentf("Test Name: %s", tt.name))

gandro replied:
Yes, `RemoveAll` clobbers its first argument (good point, I should mention that in the doc comment) and must not be used after calling the function - but I don't want to guarantee that the input argument equals the result (similar to e.g. how `append` works) - a future implementation might optimize memory consumption by re-allocating and shrinking the capacity of the resulting slice. In that case, the return value will be different from the input argument.
